vsintegration/src/FSharp.Editor/Telemetry/TelemetryReporter.fs

```diff
@@ -3,25 +3,115 @@
+        }
 
-    let private eventPrefix = "dotnet/fsharp/"
-    let private propPrefix = "dotnet.fsharp."
+    let internal lastSentEvents = ConcurrentDictionary<string, DateTime>()
 
-    let private getFullEventName name = eventPrefix + name
-    let private getFullPropName name = propPrefix + name
+    [<Literal>]
+    let eventPrefix = "dotnet/fsharp/"
 
-    let private createEvent name (props: (string * obj) list) =
-        let event = TelemetryEvent(getFullEventName name)
+    [<Literal>]
+    let propPrefix = "dotnet.fsharp."
 
-        props
-        |> List.map (fun (k, v) -> getFullPropName k, v)
-        |> List.iter event.Properties.Add
+    // This should always be inlined.
+    let inline createEvent name (props: (string * obj) array) =
+        let eventName = eventPrefix + name
+        let event = TelemetryEvent(eventName, TelemetrySeverity.Normal)
+
+        // TODO:
+        // We need to utilize TelemetryEvent's Correlation id, so we can track (for example) events on one document in the project.
+
+        // TODO: need to carefully review the code, since it will be a hot path when we are sending telemetry
+        // This particular approach is here to avoid alocations for properties, which is likely the case if we destructing them.
+        for prop in props do
+            event.Properties.Add(propPrefix + prop.Item1, prop.Item2)
 
         event
 
-    let reportEvent name props =
+[<Struct; NoComparison; NoEquality>]
+type TelemetryReporter private (name: string, props: (string * obj) array, stopwatch: Stopwatch) =
+
+    static member ReportSingleEvent(name, props) =
         let session = TelemetryService.DefaultSession
-        let event = createEvent name props
+        let event = TelemetryReporter.createEvent name props
         session.PostEvent event
+
+    // A naïve implementation using stopwatch and returning an IDisposable
+    // TODO: needs a careful review, since it will be a hot path when we are sending telemetry
+    static member ReportSingleEventWithDuration(name, props, ?throttlingStrategy) : IDisposable =
+        let throttlingStrategy =
+            defaultArg throttlingStrategy TelemetryThrottlingStrategy.Default
+
+        match throttlingStrategy with
+        | TelemetryThrottlingStrategy.NoThrottling ->
+            let stopwatch = Stopwatch()
+            stopwatch.Start()
+            new TelemetryReporter(name, props, stopwatch)
+        | TelemetryThrottlingStrategy.Throttle s ->
+            // This is not "atomic" for now, theoretically multiple threads can send the event as for now.
+            match TelemetryReporter.lastSentEvents.TryGetValue(name) with
+            | false, lastSent
+            | true, lastSent when lastSent + s.Timeout < DateTime.UtcNow ->
+                let stopwatch = Stopwatch()
+                stopwatch.Start()
+                new TelemetryReporter(name, props, stopwatch)
+            | _ -> TelemetryReporter.noopDisposable
```

Comment:
In this case we're discarding this measurement right? Would be nice to keep it and then report the average duration + count at the end.

Reply:
I wanted to implement aggregation later, not now.